### ISSUE TITLE
Fonts API refine

### ISSUE THREAD

**Issue by puzrin** [lvgl/lvgl#1006]
Reserved for signatures.

**Goals**

- Prepare API for new font features, required for good proportional fonts rendering
  - "good rendering quality" - main checkpoint.
  - "allow compact size for big fonts" - desired, but optional for 6.0
- Isolate "public" and "internal" layers.
  - Public layer can be hook-able, for example to add FreeType somehow...
  - Internal layer should not be affected by "imaginated" extensions.

**Desired features**

Minimal set is floating bitmaps + kerning + 3bpp. More advanced things about unicode are not needed right now.

As dependant - new font format required.

---

This part is only an example.

## "public" layer

Should hide implementation details from other LVGL components. Ideally, this should be transparent to rendering engine. Minimal fraction is "display string". Slower parts can not be rendered properly.

Methods:

- Init font.
- Write string to cropped area.
- Get font metrics (maxY, minY to check intersections)
- (?) Get string length in pixels.

NOTE: if we need speak seriously about replace engine with FreeType, we should consider API as "LVGL font engine", not as set of assorted functions and hooks.

## "internal" methods

This is to hide specifics like compression and storage formats from upper level.

- Draw glyph - Should simplify "drawString" implementation kerning support requires to pass correction value. At first glance, passing pair of chars looks attractive, but this can cause problem with combined chars. Seems raw kern value will be preferable
- Get glyph bitmap - hide decompression details. Return struct instead of raw buffer, bpp value, support shared buffer, ...

**Comment by puzrin** [comment]
General principles were described in first post. Here are only technical parts:

**getGlythBitmap** should provide more info

- "bounding box"
- bpp (to allow implement bpp3 via bpp4 transparently)
- should use bits-stream internally, to be ready for compression

**Draw single glyph**

- should accept kerning correction
- (?) should return cursor shift (now `advanceWidth`)

**Shared bitmap buffer for `getGlythBitmap`**

When multiple big fonts used, having separate buffer in each font can be a problem.

**Font init/structs**

- According to described 2-level API, remove ALL hooks from low levels to keep font data isolatad from code.
- If some flexibility still required, consider create font descriptors in RAM.
- Consider access font at low lovel as binary data directly, not via autogenerated code.

**Comment by puzrin** [comment]
Assorted:

- `style.text.letter_space` - probably, this should be removed. Letter spacing is complicated process, and adding fixed pixels everywhere is very rough. Suggest use different font instead. Fonts are already designed to be good. Don't try to improve those with kludges.

**Comment by kisvegabor** [comment]
Here is my suggestion based on your ideas.

**Font API**
The goal is to make it very simple and let as much work as possible to LittlevGL. Note that it's more common to store a font in external flash then using FreeType so simple "glyph level" usage should be well supported. 
Actually, it seems to me we can "make it work" with three functions (similarly to the current callbacks) but they need to be improved:
1. `get_bitmap`
  - Get the whole bitmap of the font with `bpp=1/2/4/8` in a way as LittlevGL can digest it. E.g. byte aligned line ends like now.
  - It can hide compression.
  - Get the bounding box: box_width (advanceWidth?), box_height, offset_x, offset_y
  - Get the bpp
  - It will leave the drawing to LittlevGL which makes the users' task's much easier
  - possible signature: `font->get_bitmap(font, letter, &glyph_info)`
2. `get_width`
  - Just tell the box_width
  - Possible signature: `font->get_width(font,  letter)`
3. `get_kerning`
  - An integer number should be used for performance reasons
  - However, the values can be upscaled e.g. by 16 to allow LittlveGL to sum the errors of consecutive letters.
  - Possible signature: `font->get_kerning(font, letter)`

**LittlevGL's API**
These functions are part of LittlevGL but might be replaced (e.g. for FreeType).
1. `lv_txt_get_width` 
  - Get a string's width in pixels. 
  - It's already implemented but needs to be improved
2. `lv_draw_label`
  - The core text drawing function
  - Probably rather `lv_draw_label_line` should be extracted to let LittlevGL handle text-align and other things
3. `lv_txt_letter_replace` 
  - Not implemented yet
  - FreeType can handle it in the previous two functions but if FreeType is not used LittlevGL's text shaping can come here.
  - It can be extended by the user to support specific languages.
  - It's also a good point of contribution because people can add support to their own language.

**Comment by puzrin** [comment]
Signatures looks nice.

Couple of notes:

- Term `Bounding Box`
  - When i use this in context of unpacked bitmaps, i mean bounding box of bitmap only. For example, bbox of "`,`" will be ~ 2x2 pixels.
  - When we speak about font's box, it's `height` * `advanceWidth`.
- Seems missed `lv_font_init()`.

---

> Note that it's more common to store a font in external flash then using FreeType so simple "glyph level" usage should be well supported.

If people wish to do strange things, that should not be our problem. Suggest them override `lv_draw_label` instead and leave "Font API" exclusive to LVGL.

**Comment by kisvegabor** [comment]
> It's needed to calculate position of next glyph. Drawing glyph consists of 2 part:
Draw bitmap image somewhere (bitmap bbox needed here).
Calculate cursor shift ("font_box" used).

Clear now, thank you. In this case we need:
- box width and height (returned by `font->get_bitmap`)
- y_offset, x_offset (returned by `font->get_bitmap`)
- advance_width, or simple width (returned by `font->get_width`)

Do you agree?

> Such logic may have hidden requirement to make struct const and apply additional limitations to font format (to be text only). That's not correct at architecture design stage. Requirements should be based on data first.

The task of `lv_font_init` is to create an `lv_font_t` structure for LittelvGL and depends on the actual font format  (text, bin, FreeType,  external flash, etc.).  It might have some parameters or not. 

> I don't suggest to remove FreeType support, i suggest move it to another place, where it can not make problems for architecture decisions.

The hookable `lv_txt_get_width` and `lv_draw_label` are fine to integrat FreeType, isn't it?

> Current implementation is not correct for proportional fonts (and will never be). Having this feature (kludge) can block something more useful.

Deciding which feature is useful is a different question :) However, I definitely don't want to remove a feature without a good reason. It won't take to much time to leave the support of letter spacing. Maybe it won't be perfect but still much better than nothing. 

>Problem is not with additive errors. We should calculate SUM(advanceWidth, kerning) for each separate pair. Both components can have any fractional part. And rounding can go both up and down.

It implies that advanceWidth and kerning should be stored as fractional (or upscaled).

**Comment by puzrin** [comment]
> Clear now, thank you. In this case we need:
> 
>     * box width and height (returned by `font->get_bitmap`)
> 
>     * y_offset, x_offset (returned by `font->get_bitmap`)
> 
>     * advance_width, or simple width (returned by `font->get_width`)
> 
> 
> Do you agree?

I think yes. I would include here `bpp` too.

> > Such logic may have hidden requirement to make struct const and apply additional limitations to font format (to be text only). That's not correct at architecture design stage. Requirements should be based on data first.
> 
> The task of `lv_font_init` is to create an `lv_font_t` structure for LittelvGL and depends on the actual font format (text, bin, FreeType, external flash, etc.). It might have some parameters or not.

In this case we should check all known variants and provide universal signature. Also note, `lv_font_t` is specific for LVGL engine only. Other engine can be without it or use different struct.

As far as i understand, `lv_txt_get_width` and `lv_draw_label` need some king of "font ID" + optional font state store. We should not trust current `lv_font_t` as well-designed and reflecting new needs.

I'd describe `lv_font_init` more wide:

- It should register new font under certains name/id, to be accessible from skins
- It should compile data if needed (that may include setting internal hooks/cache for more effective data handling). On more big systems that may also include precompile bitmaps from vectors.

> > I don't suggest to remove FreeType support, i suggest move it to another place, where it can not make problems for architecture decisions.
> 
> The hookable `lv_txt_get_width` and `lv_draw_label` are fine to integrat FreeType, isn't it?

Yes. In other words - "public api layer"

> > Current implementation is not correct for proportional fonts (and will never be). Having this feature (kludge) can block something more useful.
> 
> Deciding which feature is useful is a different question :) However, I definitely don't want to remove a feature without a good reason. It won't take to much time to leave the support of letter spacing. Maybe it won't be perfect but still much better than nothing.

Well, i'd suggest to mark this feature as "should not be used in architecture decisions" and move next. That will be enough. I don't care if this signature exists or not.

> > Problem is not with additive errors. We should calculate SUM(advanceWidth, kerning) for each separate pair. Both components can have any fractional part. And rounding can go both up and down.
> 
> It implies that advanceWidth and kerning should be stored as fractional (or upscaled).

Yes. And that's reflected in new format spec.

**Comment by kisvegabor** [comment]
About `lv_font_init` (the rest seems clear):
I consider `lv_font_t` as a format which is easily digestible by LittlevGL. It has callbacks to give you flexibility. So you can do any "wild" things if finally results in a `lv_font_t` struct:
- store fonts in external flash and read then on demand
- store the whole font as binary
- store the glyphs compressed
- use a TTF rendered (not necessarily FreeType but something more lighweighted) 
- etc

All these are possible scenarios which need to be supported. 

`lv_font_t` is surely not suitable for all the planned features in its current form but it doesn't seem difficult to improve it to support boundig box, advance width, kerning etc.

Once we agree on the specification and the API, I can create a new branch and manually crate a font with a few glyphs and the listed feature to see it in the practice.

**Comment by puzrin** [comment]
I think we move in right direction. Let's clarify nature of `lv_font_t` and try to drop all parts which are not mandatory there (specific to implementation).

IMHO at top level we can consider API as "font rendering engine" (with `lv_draw_label` and `lv_txt_get_width` methods). Everything else is hidden behind implementation. So, we can have:

- "LVGL string rendering engine"
- "FreeType string rendering engine"

What should be `lv_font_t` in this scope? I think, the only goal of this item - identify font to work with. This part is mandatory, everything else - not. In theory `lv_font_t` can be enum, but that's not convenient to extend. To be more precise - goal of any extention is static memory allocation for font renderer implementation.

How can we do `lv_font_t` universal and avoid pinning there specifics of "LVGL string rendering engine"? I guess - it should have minimal possible set of fields, and add the rest via `extention` field.

In terms of object programming - create base class, and inherit child classes for LVGL or for FreeType. I don't know how to do it properly in plain C, but principle is the same.

In this case you will have guarantee, that future experimenting with FreeType will not be limited by tight-coupled code. And in reverse, need of Freetype support will not make us problem in LVGL engine implementation.

---

By the way, probably it would be convenient "officially announce" term like "string renderer engine". This could be a minimal fraction unit to replace when behaviour change  needed.

For example, if anyone wish to store fonts in files, he can copy LVGL font engine source and modify for custom needs. That will be more simple than design low-level hooks in LVGL font engine. At current point low-level hooks will be very fragile. We should stabilise top level API first and make sure there are no tight-coupled things.

**Comment by kisvegabor** [comment]
I agree with the listed things. 

`lv_font_t` should contain only a very generic core and a renderer specific, changable part.

I also agree with the generalization of the renderer engine.

Give me some time to think about it.

**Comment by puzrin** [comment]
I'd like to understand one more thing. Why people may wish to use alternate engine? May this be caused by not satisfactory quality of LVGL 5.x renderer and tools? IMO if hardware has power to run FreeType, there should be space for more powerful linux GUIs.

Could you describe real use cases? I doubt here are lot of users with arabic languages and RTL need on stm32. Most of FreeType requests in issues looks like "let's add cool technology just for lulz" :).

**Comment by kisvegabor** [comment]
Very good question. Regarding FreeType, I also can't imagine a real life project where there is space for FreeType and there is no better option than LittelvGL.

However, there are tiny TTF renderer engines targeting embedded devices. Obviously, they have less feature then FreeType but there can be uses-cases where users might want real-time TTF rendering to easily change and scale fonts.

If we accept that FreeType is not a real option (or used just for fun), we can simplify a lot of things.

I'll open a new issue to ask people about it.

**Comment by puzrin** [comment]
Simple TTF rederer is not a big problem, but it requires a lot of RAM to build bitmaps atlas for fast work. Such hardware drifts to linux-based boards.

**Comment by kisvegabor** [comment]
A lot of RAM is possible. Many MCUs has external RAM interface and dev boards are equipped with SRAM. 
However, once the user generates all the fonts into RAM, it's the same if the font would be pre-generated and loaded into flash. Just the pointers in `lv_font_t` need to be adjusted.

**Comment by puzrin** [comment]
> A lot of RAM is possible. Many MCUs has external RAM interface and dev boards are equipped with SRAM.

Track 2 params: price + board size/complexity. Take Raspbery PI Zero as trigger.

If you take stm32 line prices - all SOCs with external RAM are more expensive and more complicated than raspberry.

The only exclusion could be external Quad-SPI. But minimal supporting chip is stm32F4xxx LQFP64 - also "big and expencive". 

> However, once the user generates all the fonts into RAM, it's the same if the font would be pre-generated and loaded into flash. Just the pointers in lv_font_t need to be adjusted.

Font storage format is optimized for size first, speed second. Internal caches optimized for speed first. This can be the same only for primitive cases.

**Comment by kisvegabor** [comment]
> If you take stm32 line prices - all SOCs with external RAM are more expensive and more complicated than raspberry.

I'm sure that MCU+SRAM is a viable solution and they have a lot of advantages. Just to mention a few of them which suddenly came to my mind: 
- Fast boot
- Real-time behavior
- Simple hardware
- None-complex FW

Anyway, they are in the market for a reason so we should deal with this type of configuration.

 > Font storage format is optimized for size first, speed second. Internal caches optimized for speed first. 

I agree, however you have the opportunity of customization.

> This can be the same only for primitive cases.

What "primitive cases" do you mean?

**Comment by puzrin** [comment]
> I'm sure that MCU+SRAM is a viable solution and they have a lot of advantages.

I'm not against such solutions. I explain borders from several points of view (price / complexity / tools availability), in context of GUI build.

> What "primitive cases" do you mean?

lvgl 5 format is primitive. Try to read and understand full OpenType font spec, for compare.

**Comment by kisvegabor** [comment]
> lvgl 5 format is primitive. Try to read and understand full OpenType font spec, for compare.

I see. True. However, embedded systems are also primitive compared to a PC or smartphone. We can not expect the same complexity.

Anyway, I think we are going into the right direction with new spec and API.

**Comment by puzrin** [comment]
Idea about shared buffer size. It's certainly not useable to define size manually. May pass additional metrics via #define from generated files.

First, split use cases:

### small size fonts

- 30px 4bpp => ~ 0.5K RAM.
- ~ 4 font sizes for MD theme => 2K max with separate buffers.

Shared mem may be welcome, but not top critical.

Checkpoint: stm32f0 line. With min requirement 128K flash, RAM will be 16K. Seems ok even for independent buffers.

### big fonts

This block is to be sure no technical problems exists. I don't know real use cases (not imaginated).

---

Since C does not allow load binaries directly like Rust, we will have to convert fonts into some text form (base 64, structs - not principal for this case).

Then we can add to each file something like this:

```
#define LV_ROBOTO_REGULAR_16_FONT_RECOMMENDED_GLYPH_BUFFER 754
```

Then it's up to user what to do next:

- He can ignore this info, and allow `lv_font_init()` make independent allocations
- He can use this info manually, allocate `MAX(#define values)` and pass this to `lv_font_init()`.

So, there are some manual work, but not difficult and not mandatory. We allow implement aggressive optimizations on demand, but don't force everyone pay for it.

**Comment by kisvegabor** [comment]
Speaking about the built-in font storage formats, it's a question only with compressed fonts, right?

I like the idea with `define` but in this case, the question is how to collect all the defines in `MAX(#define values)`. I mean if the user adds a new font then it's buffer size need to taken into account as well. 

If the size of the buffer would be part of `lv_font_t` it could be allocated as a local variable and passed to `bitmap_get`. E.g.
```c
uint8_t bitmap[font->buf_size];
font->bitmap_get(font, bitmap, letter, ...);
```

**Comment by puzrin** [comment]
> Speaking about the built-in font storage formats, it's a question only with compressed fonts, right?

IMHO output buffer use should be unified and should not depend on storage details. 
"Uncompressed" font may not have lines align in flash. So it still may need "output buffer". 

> I like the idea with `define` but in this case, the question is how to collect all the defines in `MAX(#define values)`. I mean if the user adds a new font then it's buffer size need to taken into account as well.

- User should write that manually, or generate via script (or track correctness via CI tests)
- If font list changed, update required. But this happens very rare.

I think this is acceptable trade-off. I don't like to pay with API complexity for not critical things.

**Comment by kisvegabor** [comment]
I have an idea! Don't allocate the buffer by default but have only a pointer to a common buffer which size will be (re)allocated on demand. Reallocate only if its size is not enough for the current glyph. This way the buffer will grow to be suitable for any glyphs.  

Reallocation will happen only when the first texts are displayed (or maybe few later) so it doesn't have any real performance drawbacks. And it eliminates all the complications and possible mistakes.

**Comment by puzrin** [comment]
I thought reallocation can cause undesired memory fragmentation.

**Comment by kisvegabor** [comment]
I don't think fragmentation is a big issue:
1. LittlevGL automatically joins free areas
2. In case of glyph buffer, big memories will be freed which can be easily reused by objects, texts, etc.

**Comment by puzrin** [comment]
If you reallocate growing memory, new continious memory block required each time. In theory, this can end with situation when you have a lot of "small blocks", but no big blocks.

> 1. LittlevGL automatically joins free areas

As far as i know, such type of garbage collector require use of indirect pointers. Does LVGL use those? If yes - then no questions, reallocation will be best choice.

**Comment by kisvegabor** [comment]
LittlevGL not uses indirect pointers but automatically combines adjacent free block into a bigger. E.g.:
- 12 byte free
- 8 byte free
- 20 byte used

will be combined into:
- 20 bytes free
- 20 bytes used 

It is proven to be able to prevent critical memory fragmentation. 

In LittlevGL dynamic memory is used very intensively so gaps are filled fast. So I think fragmentation won't be an issue.

**Comment by puzrin** [comment]
Ok, i think auto reallocate is good for start (at least). It does not require public API change and will not restrict future updates. It will also not force to store buffer size in font data.

**Comment by puzrin** [comment]
@kisvegabor do you have time to post a list of function signatures for new API (public methods + internal LVLG methods)?

That should help:

- Verify nothing missed
- Start to think if we need text format for font or can go with "binary" one.

When i say "binary" i don't mean all data should be extracted via "ReadWord(prt, offset)". Critical binary data can be partially marshalled directly to struct, without additional operations (font header, for example).

I think i can start with binary writer next week. That will be required anyway, to experiment with compression options.

**Comment by kisvegabor** [comment]
I'd rather create a new branch and manually modify a font with 3 letters to have:
- bounding box
- advance width
- kerning

This we will see it in the practice. (Smaller chance to miss anything) Do you agree?

**Comment by puzrin** [comment]
As you wish (i thought writing dummy signatures is faster, but my opinion is based on my skills only).

At first glance, uncovered part of public API is font descriptor (former `lv_font_t`) & init process. Things like bbox/adwanceWidth/kerinng are "small technical details", not affecting public API.

In parallel, if you strongly insist on text format - it would be nice to get exact example/template of new font file, to meditate with.

**Comment by kisvegabor** [comment]
> In parallel, if you strongly insist on text format - it would be nice to get exact example/template of new font file, to meditate with.

As 0the result of this experiment, we will have a template file.

I have some questions about the kerning tables. Let's say I have a font in two parts (two ranges):
1. ASCII
2. Some extra Latin letter

If we store two kerning tables, one for ASCII and an other for Latin how will we know the space between `ÁV`? ASCII doesn't know about `Á` and Latin doesn't know about `V`.

**Comment by puzrin** [comment]
> If we store two kerning tables, one for ASCII and an other for Latin how will we know the space between `ÁV`? ASCII doesn't know about `Á` and Latin doesn't know about `V`.

That's not our problem. If font contains kerning info for pair - it will be collected. If info not exists - not collected.

If you wish to mix DIFFERENT fonts, kerning info will be collected only for pairs inside each font.

**Comment by kisvegabor** [comment]
These are not different fonts (e.g. both are Roboto 20 px) only different range. Do you think in this case it's better to have a larger sparse font?

**Comment by puzrin** [comment]
> Do you think in this case it's better to have a larger sparse font?

I don't understand WHY you decided that multiple ranges can be a problem, and why we should apply additional actions.

Font authors already cared about "how to make font look nice in all imaginated ways". Just use it. If font is bad - take a good one instead :)

**Comment by puzrin** [comment]
> These are not different fonts (e.g. both are Roboto 20 px) only different range.

If you mean, those are 2 different files, subseted from original for different languages - it's very bad idea to use those, because kerning info may be missed.

https://github.com/littlevgl/lv_font_conv/blob/dev/package.json#L44 - convertor use original Roboto font. Use original sources, and everything will be fine.

**Comment by kisvegabor** [comment]
So if you need support for 10 languages you should have one big font with the all the letters of the 10 languages? It might work I just want to understand the scenario.

**Comment by kisvegabor** [comment]
Just be sure:
- ASCII subset has kerning info for letters in 0x20-0x7F range.
- Latin sunset has kerning info for letters in e.g. 0x100-0x1FF range.

So no kerning data for the e.g. 0x62 - 0x123 pair.

**Comment by puzrin** [comment]
I think, authors initially do "one big font with all letters", with proper support for all regular/expected combinations.

**Comment by puzrin** [comment]
Dumped `ĀV` chars from Roboto:

```sh
$ env DEBUG=* ./lv_font_conv.js --font ./node_modules/roboto-fontface/fonts/roboto/Roboto-Regular.woff --symbols ĀV --bpp 3 --size 40 --format dump -o 1111
```

font_info.json:

```json
{
  "ascent": 37,
  "descent": -10,
  "size": 40,
  "glyphs": [
    {
      "code": 86,
      "advanceWidth": 25.44921875,
      "bbox": {
        "x": 0,
        "y": 0,
        "width": 25,
        "height": 29
      },
      "kerning": {
        "256": -1.46484375
      }
    },
    {
      "code": 256,
      "advanceWidth": 26.09375,
      "bbox": {
        "x": 0,
        "y": 0,
        "width": 26,
        "height": 35
      },
      "kerning": {
        "86": -1.69921875
      }
    }
  ]
}
```

Kerning info exists, as expected.

**Comment by kisvegabor** [comment]
```c
      "code": 86,
      "kerning": {
        "256": -1.46484375
      }
```
Means reduce the space with 1.4 if "86" followed by "256" or "256" followed by "86"?

**Comment by puzrin** [comment]
> Means reduce the space with 1.4 if "86" followed by "256" or "256" followed by "86"?

https://github.com/littlevgl/lv_font_conv/blob/ad6557e3bac98ab1d60fee9e3a662392bd5182da/lib/collect_font_data.js#L163-L172

As far as i remember, hash contains corrections relative to "previous" glyphs =>  "256" followed by "86"

**Comment by kisvegabor** [comment]
Ok, thank you.

**Comment by kisvegabor** [comment]
I pushed the current state of the new API to the `font` branch. Here is what I've done so far:
- Reworked [lv_font_t](https://github.com/littlevgl/lvgl/blob/font/src/lv_misc/lv_font.h#L36)
- Manually reworked [lv_font_dejavu_20](https://github.com/littlevgl/lvgl/blob/font/src/lv_fonts/lv_font_dejavu_20.c) (1bpp)
- Updated the drawing and other algorithms accordingly. (It's working well)
- No kerning yet

I realized that dealing with fractional advanced width make think quite complicated in some places. We should reconsider if it's really required. 

@puzrin 
Can you send a `font_info.json:` with 0x20-0x7F range, 16px and 24 px height. I'd be curious about the advance width and kerning values to see much is the difference with and without the fractional parts.

**Comment by puzrin** [comment]
Your manually reworked font looks very far from spec. Is it something temporary, for intermediate iterations? 

> I realized that dealing with fractional advanced width make think quite complicated in some places. We should reconsider if it's really required.

That's like with kerning, some people think it may be not required :)
Could you explain exact places, where fractional part become a problem? It may be not convenient, if used everywhere, but only for width - should not be difficult.

> Can you send a font_info.json: with 0x20-0x7F range, 16px and 24 px height. I'd be curious about the advance width and kerning values to see much is the difference with and without the fractional parts.

[font_info.zip](https://github.com/littlevgl/lvgl/files/3113124/font_info.zip)

4 bits in fractional part is optimal for all cases, i checked when prepared spec.

**Comment by embeddedt** [comment]
> That's like with kerning, some people think it may be not required :)

I don't think it should be a requirement to use kerning. As we saw with the ISO-8859-1 incident, enforcing new requirements on users didn't work so well. :confused:

**Comment by puzrin** [comment]
Comments about API

1.

https://github.com/littlevgl/lvgl/blob/5321b9b62e32f80410ff3d6cd6e31d8437bdb0ce/src/lv_misc/lv_font.h#L39-L40

Nobody works with fractionals such way. Those are stored in single int/uint, and shifted in the end.

2. `lv_font_add` & `lv_font_remove` should be dropped.

3.

https://github.com/littlevgl/lvgl/blob/5321b9b62e32f80410ff3d6cd6e31d8437bdb0ce/src/lv_misc/lv_font.h#L116

Look suspicious. Nobody need bitmap only. Always with other glyph info.

4. `lv_font_get_width`, `lv_font_get_real_width` - why can't we use single function, returning right widht?

5. `lv_font_get_bpp` - this should already exist in glyph bitmap info

**Comment by kisvegabor** [comment]
1) Ok, we can combine them.

2) `lv_font_add/remove` are used to concatenate two fonts with the same height. In some cases, they can be replaced by "sparse" fonts but it can concatenate different font families too. The most typical case is to add FontAwesome symbols to normal "text" fonts. It makes possible to create labels with text like: `LV_SYMBOL_OK "Apply"` or just set `LV_SYMBOL_CLOSE` for a close button of a window without creating a new style for it. So it's very useful. The only requirement is the fonts must have the same height. In our case, the line height should be the same. However, I see that we don't have direct control over it. Anyway, the glyphs from fonts with different line-height can be arranged to the same baseline and the height of the line can be calculated according to the biggest line height. 

3) In some cases, it might take more time to get the whole bitmap than the descriptor only. Or the descriptors can be buffered on start-up because they don't need so much memory. So it increases flexibility and some people might take advantage of it.

4)  They were used to support `monospace` rendered fonts if the font is not monospaced. The same effect can be achieved by settings all the `advance width` properties to the same value. 
A use case for this "monospace trick": On the main screen of one of my projects, there is a digital clock. I set the 10 numbers to be monospaced because it looked ugly when the clock was "shaking" when a minute elapsed. Or an other one: I displayed 32 byte hexadecimal number for the user. It was good to group them like `1234 5678 ...`. It was easier to read when the groups had the same width. 
Anyway, I'm not sure it's really required because one can do it by hacking "advanced width" if something similar is required.

5) Do you mean it should be stored for every glyphs separately?

**Comment by puzrin** [comment]
(2) was attribute of old format and old convertor without merge support. It's not needed now. We don't need to propagate old approaches to new situation. New convertor allow create font with all requires chars, instead of hacking on the fly.

>  3. In some cases, it might take more time to get the whole bitmap than the descriptor only. Or the descriptors can be buffered on start-up because they don't need so much memory. So it increases flexibility and some people might take advantage of it.

I'd like to keep things simple,  something like `get_glyph_info(codePoint, withBitmap)`

If you know cases, when performance will be lost, and difference will be visible, i'd like to understand those.

>  4. ... Anyway, I'm not sure it's really required because one can do it by hacking "advanced width" if something similar is required.

I mean exactly the same. This should be hidden behind `get_width()` or `get_glyph_info()` implementations, without exposing additional methods in api. New font format has all info to implement this.

> 5\. Do you mean it should be stored for every glyphs separately?

Data format is already designed and good enougth IMHO. When we design API, goal is to draw optimal set of functions. As far as i understand, `bpp` is needed only in combination with bitmap. So, it can be returned by `get_glyph_info()`

---

I'd juggest to postpone coding, because it's like "design from bottom to top". It would be more effective focus on create header file for new font API (structs + functions). It will be more easy to discuss all at once via commit comments. It worth to commit it as separate file, to see all lines at once and be able to comment each.

**Comment by kisvegabor** [comment]
> (2) was attribute of old format and old convertor without merge support. It's not needed now. We don't need to propagate old approaches to new situation. New convertor allow create font with all requires chars, instead of hacking on the fly.

Will it be able to merge two fonts as well? E.g. Roboto and FontAwesome?

> I'd like to keep things simple, something like get_glyph_info(codePoint, withBitmap)
>If you know cases, when performance will be lost, and difference will be visible, i'd like to understand those.

As I mentioned the `dsc` can be buffered on startup. E.g. width is required more often than the bitmap so it's reasonable to do so. 

Or in the case of "fonts stored in an external flash", it's also reasonable to separate dsc from bitmap to lower the amount of data to read. 

> I mean exactly the same. This should be hidden behind get_width() or get_glyph_info() implementations, without exposing additional methods in api. New font format has all info to implement this.

Ok, I'll remove it.

> Data format is already designed and good enougth IMHO. When we design API, goal is to draw optimal set of functions. As far as i understand, bpp is needed only in combination with bitmap. So, it can be returned by get_glyph_info()

As bpp is related to the whole font why store these repetitive data in each glyph?

**Comment by puzrin** [comment]
> Will it be able to merge two fonts as well? E.g. Roboto and FontAwesome?

Sure. That was one of obvious requirements

> As I mentioned the `dsc` can be buffered on startup. E.g. width is required more often than the bitmap so it's reasonable to do so.
> 
> Or in the case of "fonts stored in an external flash", it's also reasonable to separate dsc from bitmap to lower the amount of data to read.

IMHO this can be done transparent for user. But i suggest to wait second iteration of `new_font_api.h` and check again.


> As bpp is related to the whole font why store these repetitive data in each glyph?

I don't suggest to store it in each glyph. I suggest to RETURN it as pars of glyph info. API methods should reflect demands of data flow. If bpp is used with bitmap, it's logical to return those together.

**Comment by kisvegabor** [comment]
> Sure. That was one of obvious requirements

Great! And will they have the same line-height?

> I don't suggest to store it in each glyph. I suggest to RETURN it as pars of glyph info. API methods should reflect demands of data flow. If bpp is used with bitmap, it's logical to return those together.

I see your point. It's also logical. 

I prefer keeping bpp in the font itself because it's easier to handle in the practice. Now, in `get_gsc` we can simply return an element of `lv_font_dsc_built_in_t` array. It would require more computation to create the `dsc` in run-time. As the vast majority of the users will use the built-in font format it's very important to make it fast and well optimized.

**Comment by puzrin** [comment]
> Great! And will they have the same line-height?

They will look as been designed. With the same baseline and normalized typoAscent.

It's possible to add other normalization options if needed.

> I prefer keeping bpp in the font itself because it's easier to handle in the practice. Now, in `get_gsc` we can simply return an element of `lv_font_dsc_built_in_t` array. It would require more computation to create the `dsc` in run-time. As the vast majority of the users will use the built-in font format it's very important to make it fast and well optimized.

I don't understand. Why it should be more expensive to return one more field (bsp) in glyph info? Why if user requests bpp via separate API call, this will be more cheap?

As far as i know, struct is just a convenient way to access memory offsets by literal names.

**Comment by kisvegabor** [comment]
> They will look as been designed. With the same baseline and normalized typoAscent.

Still not fully clear. Is it true that every font will have a `line_height` (or with other name) property which tells the maximum height required by the font? If two fonts are combined together then the higher `line_height` is used.

Have look at this: https://github.com/littlevgl/lvgl/blob/font/src/lv_misc/lv_font.c#L229
Now it's super simple to return a `dsc` directly because they are stored in an array, and only a pointer the requested element is returned. If `bpp` needed to be included there are 2 options:
1. `bpp` is still stored in the font and the `dsc` is created in run-time (in `get_dsc`) from the data stored in the mentioned array and completed with the bpp stored in the font.
2. Store the `bpp` in every element of `dsc` to make possible to return it immediately, but it requires more memory. 

As only the lowest level drawing function deals with the `bpp` I think we can have a unique function for that if it saves memory and/or increases speed. Probably the user never has to call `lv_font_get_bpp` because he has nothing to do with it.

**Comment by puzrin** [comment]
> Still not fully clear. Is it true that every font will have a `line_height` (or with other name) property which tells the maximum height required by the font? If two fonts are combined together then the higher `line_height` is used.

Let's clarify, use cases. Do you need "height" to count intersections or to align text?

- If you need `maxY` and `minY` - those are in font headers, and calculated for final result (probably those should be equal to HHead Asc/Desc, but that's a different story).
- If you need to count vertical align, first font will be used as "base" (it's typoAscent and typoDescent). But HHead Asc/Desc will be corrected.

---

Thanks for explanations, now understand what case you mean about dynamic create. I'd suggest to analyse all:

**Default case  (no cache).**

Struct with glyph description will be created on the fly anyway. Add one more field there will not change anything.

**Cached case.**

First, i'd suggest to check twice if that's really needed. Only critical chains should be optimized, not all. And only if slow. From my point of view, metadata requests are not biggest part of all process.

Second, if you have enougth RAM to cache all glyph metadata, there are no need to optimize it for one extra byte (size of property bpp).

> As only the lowest level drawing function deals with the `bpp` I think we can have a unique function for that if it saves memory and/or increases speed. Probably the user never has to call lv_font_get_bpp because he has nothing to do with it.

From this point of view you are right. In this case i'd suggest to not introduce public api-s from lower abstraction levels, especially, in indirect form. When you say "user may wish to keep glyph descriptions in memory for fast access", you expect that appropriate api already exists, not conflicts and not have extra demands from architecture's point of view.

**Comment by kisvegabor** [comment]
I was thinking about if this function is required at all.  The drawing function can read `lv_font_t` (font->bpp) to simply get the bpp. This function was introduced for the case if fonts with different bpp are concatenated. In this case, to get the correct bpp the letter was also required to get the corresponding font from the "list". 

However, in light of the things in #995 this function is not required because font concatenation can be removed if multiple fonts can be stored in one "file". This way the bpp will be the same for all glyphs and can be simply returned from `lv_font_t`.

Do you agree?

**Comment by puzrin** [comment]
Ah... just remembered... we forgot about about `bpp3` support. Font can be `bpp3`, but decompressor will return bitmap as `bpp4` for convenient operations.

Bitmap "extractor" does 2 additional things:

- convert bpp3 => bpp4 if needed
- align lines by word (or byte?)

IMHO it's convenient to keep such details hidden from upper level. So, we need to pass:

- actual bpp
- line width (pixels or bytes?) after align

**Comment by puzrin** [comment]
> This function was introduced for the case if fonts with different bpp are concatenated. In this case, to get the correct bpp the letter was also required to get the corresponding font from the "list".

I can understand this logic.

In general, i'd suggest to drive function creation by data flow, not by intuition and local optimizations. If you carefully inspect input/output data of each "black box" (independent module), and create optimal set of function for it, you will be very surprised with difference from initial intents :)

For example - it's "convenient" to return bpp with btmap/metadata. Because that reflects "data flow". From this point of view, it does not depend on do you have composed font or single - bpp extraction will be part of `get_bitmap()` or `get_bitmap_metadata()` calls - still no formal need to have separate method for bpp extract.

**Comment by kisvegabor** [comment]
You are right. From the "data flow's point of view, it's really logical to make it part of the `glyph dsc`. 

However, as working with resource-constrained embedded systems for so long I'm constantly thinking about memory usage and speed.  Now, it seems the whole additional complexity would be required because of bpp which is not required for the user. So I think it's just not worth to make something more complex to gain nothing (even if it is more logical).

**Comment by puzrin** [comment]
20 years ago i started as embedded developer, and i can understand your thoughts about resource constrains. That's ok if those are added to normal design flow (from top to bottom). But not ok if design flow become reversed (from bottom to top).

IMO normal design process is when you do things from top to bottom, add extra requirements and rework until satisfied. Not when you collect assorted optimization and pray those combine themself into working system :).

**Comment by kisvegabor** [comment]
>  Not when you collect assorted optimization and pray those combine themself into working system :).

It's funny and true, but our case is not so crucial. We understand the consequences very well.

Thinking about the design flow it can be considered as next iteration, right?
1. Logic says: add bpp to `glyph_dsc`
2. Will it cause problems? It makes `dsc` more complex or increases memory usage.
3. What can we do?  Move `bpp` to more generic place (lv_font_t)
4. Will it cause problems? No. We don't lose any features or don't make anything more complicated.

**Comment by puzrin** [comment]
> 2. Will it cause problems? It makes `dsc` more complex or increases memory usage.

"more complex" and "increase memory use" is not necessary "problems" in all cases. I've tried to explain here https://github.com/littlevgl/lvgl/issues/1006#issuecomment-487308968. Also, "possible improvements" should be measured somehow. At first glance, benefit will be about zero (both about memory and speed "in total"). May be i miss something important and my estimate is not correct.

> 4. Will it cause problems? No. We don't lose any features or don't make anything more complicated.

It will - api become less convenient (signature). Or additional "useless" method can appear. That become a problem, if more serious improvements required.

**Comment by kisvegabor** [comment]
I see. Your arguments are correct but I still found mines valid as well. 

I think we understood each other's thought well I can't clearly decide what to do because both ways have advantages and disadvantages.   

Finally, I'd like to go with "bpp in font" because:
1. It's a minor thing which can be changed at any time later without breaking the API
2. IMO it won't cause less problem in the long term. (I'm sure that you don't agree with this :smile: )

**Comment by puzrin** [comment]
Question is not about that single function. I used it to explain difference between final result, when based on data-flow - result would be the same for v5.

I don't suggest to avoid optimizations. I suggest to be consecutive when possible - do clear design first, then optimize if required. Because we could find satisfying solution without breaking architecture.

**Comment by kisvegabor** [comment]
While I was writing my "why I'm right" answer, I realized that "you are right". :smile: 

I realized two things:
1. It's about copying 8 bytes. Really not a performance issue
2. If the user adds an external font library, where the `dsc`s are calculated on the fly, it's difficult to handle if more `dsc`s are asked at the same time. Where to buffer them to keep them valid while their pointer is used? Changing the prototype of `get_glyph_dsc` to 
`bool get_glyph_dsc (const struct _lv_font_struct *, lv_font_glyph_dsc_t * dsc_buf, uint32_t letter);`
could solve it.

**Comment by puzrin** [comment]
May be. I could say more, when see complete `.h` file. This helps to see full picture. With single method it's very easy to fall into local micro-opts too early.

**Comment by kisvegabor** [comment]
Updates:
- kerning support added
- [lv_font.h](https://github.com/littlevgl/lvgl/blob/font/src/lv_misc/lv_font.h) updated
- [Example font](https://github.com/littlevgl/lvgl/blob/font/src/lv_fonts/lv_font_dejavu_20.c) updated

Things to do:
- General font types + functions and LittelvGL's internal font types + functions should be separated into 2 files (Now the yare in `lv_font.c`). Still didn't find a good filename for the internal format's file.
- Box coordinate system is still with top-left origo

@puzrin 
Can you review it?

**Comment by puzrin** [comment]
Could you post lv_font.h and font files with `_new` postfix or into separate folder, to see all sources as "new" (may be to separate temporary branch). That will help a lot to comment. Also worth to make it as dummy PR, to easy track comments and updates.

After we finish review, final result can override proper locations. I expect a lot of changes, not needed in real branches.

**Comment by puzrin** [comment]
About font format - seems it does not conform bin spec at all. I expected, that you can find way to express "bin format" in C (with minor tweaks). Need to understand your intents. I don't know how to comment current font src (but can comment API).

**Comment by puzrin** [comment]
API miss high-level abstractions, functions like `display_string(area, *font, x, y, *str)` & `get_width(*font, *str, prev_char)`.

Those functions are base of everything, and we can't design lower level without those. Without high level we have no answer "why functions are optimal/convenient" and will fall into endless talks about optimizations :)

If those are in separete file - consider move those here, as interface of "lvgl font engine"

**Comment by kisvegabor** [comment]
I created a PR: #1050 

> API miss high-level abstractions, functions like display_string(area, *font, x, y, *str) & get_width(*font, *str, prev_char).

I thought when we realized that FreeType doesn't have to be fully supported we dropped the need of high-level abstraction. IMO the character based API will be enough.

> About font format - seems it does not conform bin spec at all. I expected, that you can find way to express "bin format" in C (with minor tweaks). Need to understand your intents. I don't know how to comment current font src (but can comment API).

I really didn't follow it strictly. I tried to eliminate the data which is not required in LittlevGL. For example, Ascent, Descent, Typo Ascent, Type Descent is not required here to render the fonts. Instead, only an "ultimate" line_height (content area) and a base_line is required.

--- 

I'm considering to modify `font->get_glyph_dsc` as 
```c
bool (*get_glyph_dsc)(const struct _lv_font_struct *, 
                                    lv_font_glyph_dsc_t *, 
                                    uint32_t letter, 
                                    uint32_t letter_next);
```
This way `dsc->adv_w` and `dsc->kerning` can be combined into one `width`. It would provide more flexibility because we don't need to decide how to store the kerning info in the `dsc`.
In other words on a higher level it doesn't matter how  the width was calculated.

**Comment by puzrin** [comment]
> I thought when we realized that FreeType doesn't have to be fully supported we dropped the need of high-level abstraction. IMO the character based API will be enough.

I don't agree. FreeType is really not needed, but it's a good detector of good/bad split. It's better, when API is self-sufficisnt, and can solve some standalone task (=> well isolated). In our case "consistent" thing is drawing string. It still worth move high-level functions here, that should not cost too much.

> I really didn't follow it strictly. I tried to eliminate the data which is not required in LittlevGL. For example, Ascent, Descent, Typo Ascent, Type Descent is not required here to render the fonts. Instead, only an "ultimate" line_height (content area) and a base_line is required.

First. "Line Heigth" will depend on biggest glyph top/bottom. That may work for mutiline texts, but will cause labels shift is existing design if you add new big glyph into font. There are no useless data in font. Don't drop anything, at least at start.

Second, example with linear properties is the least problem, because not about data structure.

- I don't see `cmap` table with subtables. Current implementation is for array and i see no way for tuples and combinations of `format_0` + `format_sparse` (to save size).
- I don't see internal ID's used to save space.

Binary format is already optimized about size and features. If you plan to do completely different thing, we will have to recheck size again.

---

> This way `dsc->adv_w` and `dsc->kerning` can be combined into one `width`. It would provide more flexibility because we don't need to decide how to store the kerning info in the `dsc`.
> In other words on a higher level it doesn't matter how the width was calculated.

Interesting. Depends on how this will be used from upper level. But i like general idea to return kerning value (in some form) together with other data.

**Comment by kisvegabor** [comment]
> I don't agree. FreeType is really not needed, but it's a good detector of good/bad split. It's better, when API is self-sufficisnt, and can solve some standalone task (=> well isolated). In our case "consistent" thing is drawing string. It still worth move high-level functions here, that should not cost too much.

The issue here is LittelvGL supports inline recoloring. E.g. `"This is #FF0000 red #."` will recolor the word "red". With a higher level API this (or similar features later):
1.  Need to be implemented in the new functions
2.  Will not work

So if there isn't a really practical reason to provide a flexible I higher level API I would like to avoid it.

> First. "Line Heigth" will depend on biggest glyph top/bottom. That may work for mutiline texts, but will cause labels shift is existing design if you add new big glyph into font. There are no useless data in font. Don't drop anything, at least at start.

Line height should be the maximum height of the text considering every letter even if they do not exist in the font. For example, with a 20 px Roboto font containing only a `'.'` character the line height should be ~ 25 px (just guessed but not 3 px). 
Maybe is comes from `typo ascent + (abs)typo descent`?

In general, the final internal font format should be as simple as possible with only the really required information. It's crucial to make possible for users to attach new font engines. For example, I'm sure having 2 ascent and 2 descent would puzzle people. 

> - I don't see cmap table with subtables. Current implementation is for array and i see no way for tuples and combinations of format_0 + format_sparse (to save size).
>
> - I don't see internal ID's used to save space.
>
>Binary format is already optimized about size and features. If you plan to do completely different thing, we will have to recheck size again.

According to [this comment](https://github.com/littlevgl/lv_font_conv/issues/8#issuecomment-485289725) a 26 px Roboto font for ASCII is  6880 byte.

Now with the simple `const uint16_t * unicode_list;` Storing 127-32 = 95 characters needs 190 bytes. It's 2% of the whole font. Probably we could reduce it to ~50 bytes (depending on the real ranges). Then it would be 1 %. I think, for now, it's not worth the additional complexity for so little gain.

> Interesting. Depends on how this will be used from upper level. But i like general idea to return kerning value (in some form) together with other data.

Do you mean you like the idea of hiding the kerning?
I've just realized that it will hide fractional numbers too.

Let's warm-up the font concatenation topic:
We discussed earlier that font concatenation (via `font->next_page`) might be dropped because the font converter will support it. But how will it look like in the CLI? Can you provide a simple example to join Roboto 0x20-0x7F and some random FontAwesome glyphs?

**Comment by puzrin** [comment]
> So if there isn't a really practical reason to provide a flexible I higher level API I would like to avoid it.

Probably, i could not explain well. Difference between API and list of assorted functions - API is consistent and can solve some useful task (draw text string in our case). Current set of functions is not consistent. It's impossible to discuss it, because of no borders, seriously.

> The issue here is LittelvGL supports inline recoloring. E.g. "This is #FF0000 red #." will recolor the word "red".

I remember similar arguments, "someone may wish to display glyph via FreeType". Just drop this kludge and stop mixing functionality :).

If you need inline markup, create a proxy wrapper, which split string into tokens, and display each slice via "display_normal_string(*str, color)" in font engine. All that's possible with clean architecture. In other words - styling and markup is _different_ abstraction layer, not related to plain string display.

Regarding to markup example - i'd suggest to not include such primitive solutions into core, and post it into examples as plugin.

> In general, the final internal font format should be as simple as possible with only the really required information. It's crucial to make possible for users to attach new font engines. For example, I'm sure having 2 ascent and 2 descent would puzzle people.

Try to imagine simple scenario - you decide to calculate line_height different, and required data not exists... Ups... and you have to search @puzrin and ask him update font convertor... :).

I know a bit more situations, when additional metrics can be used. Saving 12 flash bytes does not worth extra risks.

> According to [this comment](https://github.com/littlevgl/lv_font_conv/issues/8#issuecomment-485289725) a 26 px Roboto font for ASCII is 6880 byte.
> 
> Now with the simple `const uint16_t * unicode_list;` Storing 127-32 = 95 characters needs 190 bytes. It's 2% of the whole font. Probably we could reduce it to ~50 bytes (depending on the real ranges). Then it would be 1 %. I think, for now, it's not worth the additional complexity for so little gain.

Advandage of bin spec is that it's optimal for ALL situations, not only for `ascii 26px Roboto 3bpp`. Bitonal 12px without kerning takes 1300bytes only. Also, your current solution does not supports codes > 0xFFFF (emojis, for example), and sparse ranges (may be you know how to do it, but that's not clear from src).

```
env DEBUG=* ./lv_font_conv.js --font ./node_modules/roboto-fontface/fonts/roboto/Roboto-Regular.woff -r 0x20-0x7f --format bin --size 12 --bpp 1 -o test.bin --no-kerning
  font last_id: 96 +0ms
  font minY: -3 +3ms
  font maxY: 10 +0ms
  font glyphIdFofmat: 0 +1ms
  font kerningFormat: 0 +0ms
  font advanceWidthFormat: 0 +0ms
  font xy_bits: 4 +0ms
  font wh_bits: 4 +1ms
  font advanceWidthBits: 5 +0ms
  font monospaced: false +0ms
  font indexToLocFormat: 0 +4ms
  font.table.head table size = 44 +0ms
  font.table.cmap table size = 124 +0ms
  font.table.loca table size = 204 +0ms
  font.table.glyph table size = 928 +0ms
  font font size: 1300 +6ms
```

> Do you mean you like the idea of hiding the kerning?
> I've just realized that it will hide fractional numbers too.

I mean "probably we don't need separate method to extract kerning". No, it will not hide fracts. Joining all to "width" is not correct. Kerning is correction to left, and advanceWidth is "real width from left to right", it will be very confusing if it varies.

PS. You could add property "corrected_int_width", that will be less confusing. But i think it does not worth of efforts.

> Can you provide a simple example to join Roboto 0x20-0x7F and some random FontAwesome glyphs?

`./lv_font_conv.js --font roboto.woff -r 0x20-0x7f --font font-awesome -r 0xfe00-0xfe20 --size 16 --bpp 3 --format bin -o output.font`

**Comment by kisvegabor** [comment]
> Probably, i could not explain well. Difference between API and list of assorted functions - API is consistent and can solve some useful task (draw text string in our case). Current set of functions is not consistent. It's impossible to discuss it, because of no borders, seriously.

I see and agree with in general. But I believe we don't need to make string drawing public i.e. no interface is required for it. On the one hand, it really would be cleaner. However, does it really worth the extra work? So far, nobody asked something like this. In comparison, glyph level interface was asked several times and the current implementation serves real needs of the users.  So I just don't want to spend time on something which won't be used by people.

> Regarding to markup example - i'd suggest to not include such primitive solutions into core, and post it into examples as plugin.

Markup(s) need to be core functionality. For example, because the "text width" and "get a letter's position" should ignore the markups. 

We need to limit the scope of the font converter. The original goal was to rewrite the PHP converters in Node.JS. I understand that some things turned out while designing the new converter so I'm fine to make some updates on fonts. But redesigning, rewriting and retesting core functionalities (new API for strings drawing, rework markup, etc) are not planned features now.  I'm saying this because I have only a few hours in a day to work on LittlevGL and at that time I'd like to work on features which are often asked by people. Don't misunderstand me, I highly appreciate your efforts, and I agree this part can be improved. However, I/we need to prioritize things.

> Try to imagine simple scenario - you decide to calculate line_height different, and required data not exists... Ups... and you have to search @puzrin and ask him update font convertor... :).
I know a bit more situations, when additional metrics can be used. Saving 12 flash bytes does not worth extra risks.

So far only **glyphs** and **bitmaps** were abstracted by callback functions but the font level properties (like line_height) are not. So there is no way to gather font info (similarly to `glyph_dsc`) which would be able to hide font ascents/descents etc. If we added abstraction here as well then it could hide the internals and return the ultimate `line_height`. In summary we can have a general `lv_font_info_t` type and a `get_font_info_cb` callback to fill that struct.

> Advandage of bin spec is that it's optimal for ALL situations, not only for ascii 26px Roboto 3bpp. Bitonal 12px without kerning takes 1300bytes only. Also, your current solution does not supports codes > 0xFFFF (emojis, for example), and sparse ranges (may be you know how to do it, but that's not clear from src).

In case of a continuous font  `unicode_list = NULL` because the mapping is clear from `unicode_first` and `unicode_last`.

The current solution can support a 64k range because the Unicode values can be relative to `unicode_first`. (Not works this way yet)

> I mean "probably we don't need separate method to extract kerning". No, it will not hide fracts. Joining all to "width" is not correct. Kerning is correction to left, and advanceWidth is "real width from left to right", it will be very confusing if it varies.

What's the difference? From the outer world's point of view, the only question is how many pixels we should move to the right. If it comes from "advanced width + kerning" or just "advanced width" doesn't really matter. Or am I missing something?

> ./lv_font_conv.js --font roboto.woff -r 0x20-0x7f --font font-awesome -r 0xfe00-0xfe20 --size 16 --bpp 3 --format bin -o output.font

It's really clear and simple, thank you. 
Let's postpone this question until the others become clearer.

**Comment by embeddedt** [comment]
@kisvegabor @puzrin 

> We need to limit the scope of the font converter. 

I completely agree, especially since it's already May and we need time for people to test all the changes and inevitably find bugs/drawbacks. A lot more has been changed in v6.0 than any of the minor 5.x releases.

**Comment by puzrin** [comment]
> I see and agree with in general. But I believe we don't need to make string drawing public i.e. no interface is required for it. On the one hand, it really would be cleaner. However, does it really worth the extra work?

Problem is, that [current extensions (markup, (?)freetype) + some strange hooks & code injections into font data] are hacks, which make  code to be more and more tight-coupled. Instead of been strictly isolated. That cause extending (improvement) very difficult. If currently proposed things were done in v5, replacing fonts would be much more easy now.

> So far, nobody asked something like this. In comparison, glyph level interface was asked several times and the current implementation serves real needs of the users.

That's a philosophic question :).

People may not ask "something like this" because they don't know good alternatives. For example, people don't ask Material Design. Does it mean we should continue with "simple" interfaces only?

Also, people may ask "strange things" (IMO), like glyph-level interface, because they don't know something better and are limited with current implementation. That does not mean such requests are good and better alternatives not available.

> So I just don't want to spend time on something which won't be used by people.

I understand your constrains and respect your time. So, i never consider you as "my free programmer", and can suggest compromises how to do things "better" WITHOUT additional time expenses from your side.

- I did significant work about convertor & font spec, because understand that nobody of available collaborators can do it "well" in reasonable future. Try to count your saved time, how many iterations could you spend to reach comparable level :). I don't mean you can never do this, i ask about your [saved] time only.
- I suggested to stop coding and discuss API, prior to reject "big changes". That's relatively cheap. Even if you will not agree with "big rewrite", that will not cost too much. But we could try at least, instead of rejecting this way without investigation and continue extend old "not extendible" architecture. In worst case, result still will be useable for next iteration in v7.

Also note, you don't like to spent extra time for "api rewrite", but you spend it for new font text format development. May be new api + binary format will have comparable cost?

> In case of a continuous font `unicode_list = NULL` because the mapping is clear from `unicode_first` and `unicode_last`.
> 
> The current solution can support a 64k range because the Unicode values can be relative to `unicode_first`. (Not works this way yet)

I believe, you can invent your own working solution. But does it worth spent time for it? I suggested you a battle-tested and extendible thing. Not because "invented by me", but because carefully scraped from OpenType, with knowledge of major pitfalls. Note, i have some experience with fonts, but still prefer to not invent new standards when possible.

Instead of inventing new text format, i'd suggest to spent time for consistent API ("font engine").

> Markup(s) need to be core functionality. For example, because the "text width" and "get a letter's position" should ignore the markups.

I guess, more correct answer would be "markup(s) need to be in core with current design, due current design limitations" :)

IMO that's a road to "nowhere" with tight-coupled code. Consider as axiom _"architecture (modules isolation and abstraction layers) should never be broken"_ and try to find solution with this restrictions. It's possible. I can help with design, if needed.

> We need to limit the scope of the font converter. The original goal was to rewrite the PHP converters in Node.JS.

I do not agree with this approach. It's very limited. We do not need "just new font convertor", we need "new features". May be part of lvgl users need only "better php replacement". But my personal dreams also include "right font rendering" and "refactoring font-related code to make it really extendible". All that in context of "providing modern interfaces". I hope, such "dreams" will be useful for all, not for me only.

Initially, i started to "rewrite convertor". That's true. But when dived inside, i understood "everything is bad" :). So, i spend additional efforts with font spec design to cover all "holes" and provide some guarantees that bin format will not need more changes. 

> What's the difference? From the outer world's point of view, the only question is how many pixels we should move to the right. If it comes from "advanced width + kerning" or just "advanced width" doesn't really matter. Or am I missing something?

I'd say, it will be very confusing to use "combined" `advanceWidth`, which can vary, depending on situation. When people see "width", they expect width, not magic. At least, i suggest to give it different name, something like "cursor_shift".

Second problem is, that correction depends on previous glyph, but you tend to use next glyph (as far as i understood). That all may work, but will look very cryptic.

Adding correction data to same function call will not have side effects and can simplify api. Combining it with width - will have side effects, and saving single `>> 4` to remove fracts does not worth potential problems with code readability.

---

In short:

- I understand your intents, but those are not API rewrite in my eyes. My participation is not needed in this case (my architect skills will not be useful, and that's not interesting for me).
- Independent on your decision, i'm not going block moving to v6 release. If you decide to use that "strange" (from my point of view) text format, i'll try to make writer for it in convertor.

**Comment by puzrin** [comment]
@embeddedt 

> > We need to limit the scope of the font converter.
> 
> I completely agree, especially since it's already May and we need time for people to test all the changes and inevitably find bugs/drawbacks. A lot more has been changed in v6.0 than any of the minor 5.x releases.

I don't know release deadlines. If schedule pressure is high, fonts may be postponed at all to next release (that should not affect other components). May be with announce that v6.0 will be not LTS.  If fonts will be done at 7.0, migration 6=>7 will be easy.

**Comment by kisvegabor** [comment]
**Public API and limiting the scope**
The glyph level API is proven to be useful in several cases. It was created to serve the real needs of people. So I'd like to keep it because people like and use it.

The string drawing API is a different thing and **independent** from glyph level API. They have a different purpose and one can't replace the other. In addition, regardless of the glyph level interface the string drawing API can exist or not, they don't bother each other. 

Adding your own string drawing function is a difficult thing and requires more work from you. And the result can be only slightly better generally or much better for "exotic" languages. In comparison, the glype level interface is very simple yet powerful.

The string level drawing and string operations are really tightly coupled to the core code. Creating a public API for it would be a huge work.

As the two interfaces are independent I suggest postponing the public API. We can add it at any time if we feel that it's required. 

**Text format**
If we drop font concatenation of fonts something like `cmap` will be required instead. I will investigate it as I can. 

**Width** 
> I'd say, it will be very confusing to use "combined" advanceWidth, which can vary, depending on situation. When people see "width", they expect width, not magic. At least, i suggest to give it different name, something like "cursor_shift".

Ok, let use a different name than `width`. `cursor_shift` can work for first. 

> Second problem is, that correction depends on previous glyph, but you tend to use next glyph (as far as i understood). That all may work, but will look very cryptic.

Ups, it's my bad. I'll change it to use `prev_letter`

> Adding correction data to same function call will not have side effects and can simplify api. Combining it with width - will have side effects, and saving single >> 4 to remove fracts does not worth potential problems with code readability.

I suggested hiding it because I realized that LittlevGL sees only an integer width (cursor shift). It doesn't see the details, just gets how many pixels should it skip the draw the next letter. Therefore even the `>>`, and kerning is hidden from LittlevGL. Now `lv_font_get_width(font, letter, letter_next)` (will use `letter_prev`) is used everywhere and that's enough. If the glyph dsc had a `cursor_shift` field then it would replace this function.

**Comment by puzrin** [comment]
> **Public API and limiting the scope**
> The glyph level API is proven to be useful in several cases. It was created to serve the real needs of people. So I'd like to keep it because people like and use it.

I feel some "recursion" in this words ("let's keep hacks because we already used it in code, instead of fix them all").

May be i don't know something, but in this case i'd like to understand exactly, where such feature used. As you seen, i never suggested to drop feature at all and could provide examples of alternate solution (last one - with markup). That's a scope of architecture design process - provide good solution for all required use cases. Not to select first possible and force the rest use it via hacks.

Right now i can imagine only one place where glyph-level could be used - split multiline string to fit with in textarea. But this can be done on "string level".

> Adding your own string drawing function is a difficult thing and requires more work from you. And the result can be only slightly better generally or much better for "exotic" languages. In comparison, the glype level interface is very simple yet powerful.

In lvgl, this [string-level] code should exists anyway, so we can say, cost is ~ the same, independent of code location.

For users - they can just copy code from lvgl. If we concentrate it in isolated "font engine", that will be not difficult. And will give guarantee that external code will not be broken.

> The string level drawing and string operations are really tightly coupled to the core code. Creating a public API for it would be a huge work.

examples... please...

I understand, that's not trivial, but does not seems so difficult, if we carefully inspect all cases. In worst case, we could postpone font changes for v7.0, to not break schedule.

> As the two interfaces are independent I suggest postponing the public API. We can add it at any time if we feel that it's required.

I would not ask for "string API" if it could be isolated. In my eyes, glyph api is "wrong by design" and should be replaced, not enhanced. We could postpone all, but not string's part.

> **Text format**
> If we drop font concatenation of fonts something like `cmap` will be required instead. I will investigate it as I can.

Exactly. Concatenation is consequence of design decision to not have 'cmap' (is it good or bad - different question). You now have the same problem with kerning (impossible to add subtable format 3 support), because mix it with glyph data.

In OpenType all this is solved by "tables" approach. This also helps to extend features - other tables are not affected.

> > Adding correction data to same function call will not have side effects and can simplify api. Combining it with width - will have side effects, and saving single >> 4 to remove fracts does not worth potential problems with code readability.
> 
> I suggested hiding it because I realized that LittlevGL sees only an integer width (cursor shift). It doesn't see the details, just gets how many pixels should it skip the draw the next letter. Therefore even the `>>`, and kerning is hidden from LittlevGL. Now `lv_font_get_width(font, letter, letter_next)` (will use `letter_prev`) is used everywhere and that's enough. If the glyph dsc had a `cursor_shift` field then it would replace this function.

That return us to question of "high level string-based api". It will be normal to  have `get_width_int(string)`, and use it everywhere externally. If nobody uses glyph access directly, they will be not affected by our internal machinery.

All "problems" come from your wish to use glyph-level API. But my suggestions are based on assumption, that string-based API used. Then all those cases just not exists. I would never suggest to spread fractional numbers to all lvgl, that's really madness :)

**Comment by embeddedt** [comment]
@puzrin 
> For users - they can just copy code from lvgl. If we concentrate it in isolated "font engine", that will be not difficult.

I don't think most users will go to the trouble of looking inside the font library to see how to implement the glyph-level API. If it's not already there and waiting for them, I think most people will assume it's not supported.

**Comment by kisvegabor** [comment]
Again, when we started the whole converter thing I said "yes" to rewrite the converters in a compatible way. Now we are speaking about the redesign of the architecture. It's really not what I want. What I'm interested in right now are a better CLI font converter and some improvements in fonts which meanly means kerning and the compression of bitmaps while keeping compatibility as much as possible. 

I'm sure that a lot of things can be improved in the architecture but it's not that bad that it's worth the effort.

**Comment by puzrin** [comment]
I see, we both did not know all technical details before start. I could not rewrite it "in compatible way" :). Where do we move now? Here 2 "big" choices to do:

- Switch or not to binary format (text one is difficult to describe all required features).
- Switch or not to string-based API (improve code isolation/consistency: avoid fractional-s spread to other components and so on).

Both are not for fun, and IMO required. But there are deadlines and human resource constrains.

> which meanly means kerning and the compression of bitmaps while keeping compatibility as much as possible.

IMHO at font format level that's technically impossible, because some data types (glyph id size) are dynamic, and offsets used - that's needed for compact size.

At API level you can temporary mask unsupported features. For example, round fractional values inside or ignore some metrics. That's not good for precision, but will allow use existing code with less changes.

**Comment by kisvegabor** [comment]
> Switch or not to binary format (text one is difficult to describe all required features).

I'll try the text format first. I'm almost sure it will work.

> Switch or not to string-based API (improve code isolation/consistency: avoid fractional-s spread to other components and so on).

I don't think it's required so let's postpone it.

**Comment by puzrin** [comment]
Ok. I respect your choice. Then this issue seems not needed.

**Comment by puzrin** [comment]
Ok.

Could you consider assign migration to string api for v7 roadmap?

**Comment by puzrin** [comment]
BTW, one of "simple" ways to inject colors into strings:

- Create a set of special control codes with meaning: `COLOR1`, `COLOR2`, ... `END`.
- Set width 0 for those virtual chars.
- Override string output.

That's hacky, but without breaking things.

**Comment by kisvegabor** [comment]
> BTW, one of "simple" ways to inject colors into strings:

Sounds good at first glance.

In light of [this comment](https://github.com/littlevgl/lvgl/issues/1006#issuecomment-490491274) I close this issue. Let continue in #1057

**Comment by kisvegabor** [comment]
> Could you consider assign migration to string api for v7 roadmap?

Let's decide it when the current goals are reached :)
